File: search-algorithms/binary-search.js
/**
 * Binary Search Implementation
 *
 * Time complexity - O(log n)
 *
 * Searches the input searchItem from a sorted array
 * @param {number[]} inputArray Sorted input array
 * @param {number} searchItem Number to be searched
 * @returns {number} Index of the searchItem in inputArray
 */

const binarySearch = (inputArray, searchItem) => {
  let low = 0;
  let high = inputArray.length - 1;
  while (low <= high) {
    let mid = Math.round((low + high) / 2);
    let midTerm = inputArray[mid];
    if (midTerm === searchItem) {
      return mid;
    }
    if (midTerm < searchItem) {
      low = mid + 1;
    } else if (midTerm > searchItem) {
      high = mid - 1;
    }
  }
};

const testArray1 = [2, 4, 6, 8, 9, 10];
const testArray2 = [11, 22, 31, 35, 36, 73];

console.log(binarySearch(testArray1, 6));
console.log(binarySearch(testArray2, 36));

File: search-algorithms/binary-search.py
def binary_search(sorted_input_array, search_item):
    '''
    Binary Search Implementation

    Time complexity - O(log n)

    Searches the input searchItem from a sorted array

    Parameters
    ----------
    sorted_input_array : list of int
        Sorted input array `sorted_input_array`
    search_item : int
        Number to be searched `search_item`

    Returns
    -------
    Index of the search_item in sorted_input_array
    '''
    low = 0
    high = len(sorted_input_array)-1
    while(low <= high):
        mid = (low+high)//2
        mid_term = sorted_input_array[mid]
        if(mid_term == search_item):
            return mid
        elif(mid_term < search_item):
            low = mid+1
        elif(mid_term > search_item):
            high = mid-1


test_array1 = [2, 3, 6, 7, 8, 12, 56]
test_array2 = [12, 33, 46, 47, 58, 62, 86]

print(binary_search(test_array1, 8))
print(binary_search(test_array2, 33))
